util/dvsim/SGE.py

```diff
@@ -0,0 +1,337 @@
+            try:
+                m = int(array[1])
+            except (ValueError, IndexError):
+                m = None
+            try:
+                s = int(array[2])
+            except (ValueError, IndexError):
+                s = None
+            if m == s is None:
+                qsuboptslist.append('-t %d' % n)
+            elif s is None:
+                qsuboptslist.append('-t %d-%d' % (n, m))
+            else:
+                qsuboptslist.append('-t %d-%d:%d' % (n, m, s))
+
+        qsubopts = ' '.join(qsuboptslist)
+
+        pout = _exec(self.cmd_qsub, stdin=qsubopts + '\n' + job,
+                     print_command=False)
+
+        try:
+            # Next to last line should be
+            # "Your job 1389 (name) has been submitted"
+            # parse for job id
+            jobid = int(pout.split('\n')[-2].split()[2])
+            return jobid
+        # except (ValueErrorValueError, IndexError, AttributeError) (e):
+        except (ValueError, IndexError, AttributeError):
+            error_msg = """Error submitting SGE job:
+%s
+%s
+
+Output was:
+%s""" % (qsubopts, job, pout)
+            logger.error(error_msg)
+            raise IOError(error_msg)
+
+    def getuserjobs(self, user=pwd.getpwuid(os.getuid())[0]):
+        """Returns a list of SGE jobids run by a specific user
+
+        Inputs
+            user - SGE user to poll (Default = '', i.e. current user)
+            qstat - path to qstat binary (Default = 'qstat')
+        """
+
+        p = subprocess.Popen(self.cmd_qstat + " -u " + user, shell=True,
+                             stdout=subprocess.PIPE)
+        qstat_output, _ = p.communicate()
+        joblist = JobList(qstat_output)
+        return [job for job in joblist if job.user == user]
+
+    def run_job(self, command, name='default', logfnm='default.log',
+                wait=True):
+        """Run job on SGE with piped logging."""
+        jobid = self.submit(command, name=name, stdout=logfnm,
+                            stderr=logfnm, wait=wait)
+        return jobid
+
+    def get_queue_instance_status(self):
+        """
+        Get loads for each queue instance
+        """
+        output = _exec(' '.join([self.cmd_qstat, '-f']))
+
+        data = []
+        for line in output.split('\n')[1:]:
+            t = line.split()
+            if len(t) != 5:
+                continue
+
+            nodename = t[0].split('@')[1].split('.')[0]
+            maxslots = int(t[2].split('/')[2])
+            load = float(t[3])
+
+            data.append({'name': nodename, 'maxslots': maxslots, 'load': load})
+
+        return data
+
+
+def _exec(command, print_to_screen=False, logfnm=None, stdin='',
+          print_command=False):
+
+    """
+    Runs command line using subprocess, optionally returning stdout
+    """
+    logger = logging.getLogger('_exec')
+
+    print_to_file = (logfnm is not None)
+    f = None
+    if print_to_file:
+        try:
+            f = open(logfnm, 'a')
```

Comment:
@sha-ron @matutem in order to use the `with open(...) as f:` pattern here, we could restructure this function as follows:

```python
def _exec(command, print_to_screen=False, logfnm=None, stdin='',
          print_command=False):
    """
    Runs command line using subprocess, optionally returning stdout
    """
    def _call_cmd(command, stdin=''):
        p = subprocess.Popen(command, shell=True, stdin=subprocess.PIPE,
                             stdout=subprocess.PIPE,
                             stderr=subprocess.STDOUT)
        output, _ = p.communicate(stdin)
        return output

    logger = logging.getLogger('_exec')

    if print_command:
        logger.info("Executing process: \x1b[1;92m%-50s\x1b[0m Logfile: %s",
                    command, logfnm)

    output = ""
    if logfnm is not None:
        try:
            with open(logfnm, 'a') as f:
                if print_command:
                    print(f, "Executing process: %s" % command)
                output = _call_cmd(command, stdin)
                f.write(output)
        except IOError:
            error_msg = 'Error: File: ' + str(logfnm) + ' does not appear to exist.'
            logger.error(error_msg)
            raise IOError(error_msg)
    else:
        output = _call_cmd(command, stdin)

    logger.info('Output of command is:\n%s', output)

    if print_to_screen:
        print(output)

    return output
```

Reply:
The way the current PR opens the file will close when the "with" context goes out of scope, so in line 320 the file will be closed and you'll probably get some other exception.

Michael's suggestion extends the scope of the "with" context past the file write. Thanks Michael.

Reply:
Thanks Michel , I've fixed it

Reply:
I don't see the necessary changes. Could you upload the fix?